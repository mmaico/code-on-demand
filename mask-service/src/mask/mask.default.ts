import {Type} from "./formatter/mask.formatter";


export class MaskDefault {

    public static masks = new Map<Type, string>([
        [Type.CPF, "###.###.***-**"],
        [Type.CNPJ, "**.***.***/****-##"],
        [Type.MOBILE, "(##)**###-####"],
        [Type.EMAIL, "#######@#***.###.##"]
    ])
}