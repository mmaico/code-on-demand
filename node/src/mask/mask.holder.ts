export class MaskHolder {
    public static readonly HASH: string = "#"
    public static readonly STAR: string = "*"
    public static readonly AT: string = "@"
}