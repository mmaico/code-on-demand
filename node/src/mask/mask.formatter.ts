import {MaskHolder} from "./mask.holder";

export enum Type { GENERIC, EMAIL }

export interface Formatter {
    apply(data: string, mask: string): string
}


export class GenericFormatter implements Formatter {
    apply(data: string, mask: string): string {
        let treatedData = data.replace(/\*/g, "0").trim()
        let masked = []

        for (let controller = 0, i = 0; i < mask.length; i++) {
            let character = mask.charAt(i)
            if (character === MaskHolder.HASH) {
                masked.push(MaskHolder.getChar(treatedData, controller))
                controller++
            } else {
                masked.push(character)
                controller = character === MaskHolder.STAR ? controller + 1: controller
            }
        }

        return masked.join("")
    }

}

export class EmailFormatter implements Formatter {

    apply(email: string, mask: string): string {
        let masked = []
        let maskTreated = MaskHolder.treatEmailMask(mask)

        for (let controller = 0,  i = 0; i < email.length; i++) {
            let maskCharacter = MaskHolder.getChar(maskTreated, controller);

            if (maskCharacter == MaskHolder.HASH || email.charAt(i) == MaskHolder.AT || MaskHolder.getChar(maskTreated, controller) == email.charAt(i)) {
                if (masked.length < 4 && email.charAt(i) == MaskHolder.AT) {
                    masked.push("******" + email.charAt(i));
                    while(MaskHolder.getChar(maskTreated, controller) != MaskHolder.AT) controller++;
                } else {
                    masked.push(email.charAt(i));
                }
                controller++;
            } else {
                masked.push(MaskHolder.STAR);
                controller = maskCharacter == MaskHolder.STAR ? controller + 1 : controller;
            }
        }

        return masked.join("").trim()
    }

}