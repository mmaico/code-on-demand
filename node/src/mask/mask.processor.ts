import jsonpath = require('../jsonpath');
import {EmailFormatter, Formatter, GenericFormatter, Type} from "./mask.formatter";

export class MaskProcessor {

    private formatters: Map<Type, Formatter> = new Map<Type, Formatter>([
        [Type.GENERIC, new GenericFormatter()],
        [Type.EMAIL, new EmailFormatter()]
    ])

    process(json: any, settings: Setting[]): string {
        settings.forEach(setting => {
            let result = (jsonpath.JSONPath({json, path: setting.path, resultType: 'all'}) as Array<any>)
            result.forEach(field => {
                let type: Type = Type[setting.type]
                field.parent[field.parentProperty] = this.formatters.get(type).apply(field.value, setting.mask)
            })
        })

        return JSON.stringify(json)
    }

}

export class Setting {
    constructor(public type: string, public path: string, public mask: string) {

    }
}