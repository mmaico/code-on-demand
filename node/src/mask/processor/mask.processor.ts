import * as jsonpath from '../../../lib/jsonpath';

import {MaskDefault} from "../mask.default";
import {EmailFormatter, Formatter, GenericFormatter, Type} from "../formatter/mask.formatter";

export class MaskProcessor {

    private formatters: Map<Type, Formatter> = new Map<Type, Formatter>([
        [Type.CPF, new GenericFormatter()],
        [Type.CNPJ, new GenericFormatter()],
        [Type.MOBILE, new GenericFormatter()],
        [Type.EMAIL, new EmailFormatter()]
    ])

    process(json: any, settings: Setting[]): string {
        settings.forEach(setting => {
            let result = (new jsonpath.JSONPath({json, path: setting.path, resultType: 'all'})) as unknown as Array<any>

            result.forEach(field => {
                let type: Type = Type[setting.type]
                field.parent[field.parentProperty] = this.formatters.get(type).apply(field.value, MaskDefault.masks.get(type))
            })
        })
        return JSON.stringify(json)
    }

}

export class Setting {
    constructor(public type: string, public path: string) {

    }
}