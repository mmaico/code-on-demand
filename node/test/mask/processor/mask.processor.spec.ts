import ScenarioHelper, {scenario} from "../../infratest/scenario_helper.infra";
import {maskProcessorScenarios} from "./mask.processor.scenarios";
import {And, Given, Then, When} from "../../infratest/BDD";
import {MaskProcessor} from "../../../src/mask/processor/mask.processor";

describe('Mask processor', () => {

    beforeAll(async () => {
        await ScenarioHelper.load(maskProcessorScenarios)
    })

    it('should apply mask on cpf and mobile', () => {
        Given("A a list of sellers with cpf and phone data")
           const sellers =   scenario('sellers with cpf and mobile data')
        And('the json path settings')
            const jsonPathSettings = scenario('settings(jsonpath) to mask the cpf and mobile')
        When('the mask process is called')
           const result = new MaskProcessor().process(sellers, jsonPathSettings)
        Then('must return with masked cpf and mobile')
            expect(result).toMatchSnapshot()
    })

    it('should apply mask on cnpj and email', () => {
        Given("A a list of sellers with cnpj and email data")
            const sellers =   scenario('sellers with cnpj and email data')
        And('the json path settings')
            const jsonPathSettings = scenario('settings(jsonpath) to mask the cnpj and email')
        When('the mask process is called')
            const result = new MaskProcessor().process(sellers, jsonPathSettings)
        Then('must return with masked cnpj and email')
            expect(result).toMatchSnapshot()
    })
})