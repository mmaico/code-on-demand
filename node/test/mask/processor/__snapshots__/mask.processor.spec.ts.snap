// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mask processor should apply mask on cnpj and email 1`] = `"{\\"store\\":{\\"sellers\\":[{\\"name\\":\\"Lannisters s/a\\",\\"cnpj\\":\\"**.***.***/****-10\\",\\"mobile\\":\\"118857454\\",\\"email\\":\\"su*****@l*********.com\\"},{\\"name\\":\\"Starks s/a\\",\\"cnpj\\":\\"**.***.***/****-55\\",\\"mobile\\":\\"118849948\\",\\"email\\":\\"su*****@s*****.com\\"}]}}"`;

exports[`Mask processor should apply mask on cpf and mobile 1`] = `"{\\"store\\":{\\"sellers\\":[{\\"name\\":\\"Jaime Lannister\\",\\"cpf\\":\\"921.***.***-**\\",\\"mobile\\":\\"(11)*****-9852\\",\\"email\\":\\"jlannister@got.com\\"},{\\"name\\":\\"Sansa Stark\\",\\"cpf\\":\\"107.***.***-**\\",\\"mobile\\":\\"(11)*****-7896\\",\\"email\\":\\"sstark@got.com\\"}]}}"`;
